spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageResource.java

```diff
@@ -0,0 +1,332 @@
+import java.util.concurrent.TimeUnit;
+
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Bucket;
+import com.google.cloud.storage.BucketInfo;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageException;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.core.io.WritableResource;
+import org.springframework.util.Assert;
+
+/**
+ * Implements {@link WritableResource} for reading and writing objects in Google Cloud
+ * Storage (GCS). An instance of this class represents a handle to a bucket or a blob.
+ * A resource path that refers to a bucket must end with forward slash (/).
+ *
+ * @author Vinicius Carvalho
+ * @author Mike Eltsufin
+ * @author Chengyuan Zhao
+ * @author João André Martins
+ */
+public class GoogleStorageResource implements WritableResource {
+
+	private static final Log LOGGER = LogFactory.getLog(GoogleStorageResource.class);
+
+	private final Storage storage;
+
+	private final URI location;
+
+	private final String bucketName;
+
+	private final String blobName;
+
+	private final boolean autoCreateFiles;
+
+	/**
+	 * Constructs the resource representation of a bucket or a blob (file) in Google Cloud Storage.
+	 * Note that a location that refers to a bucket, must end with a forward slash, e.g., gs://your-bucket/.
+	 * @param storage the Google Cloud Storage client
+	 * @param location the URI of the bucket or blob, e.g., gs://your-bucket/ or gs://your-bucket/your-file-name
+	 * @param autoCreateFiles determines the auto-creation of the file in Google Cloud Storage
+	 *                              if an operation that depends on its existence is triggered
+	 *                              (e.g., getting the output stream of a file)
+	 * @throws IllegalArgumentException if the location URI is invalid
+	 */
+	public GoogleStorageResource(Storage storage, String location,
+			boolean autoCreateFiles) {
+		Assert.notNull(storage, "Storage object can not be null");
+		Assert.isTrue(location.startsWith(GoogleStorageProtocolResolver.PROTOCOL), "Location must start with gs://");
+
+		this.storage = storage;
+		this.autoCreateFiles = autoCreateFiles;
+		try {
+			this.location = new URI(location);
+			this.bucketName = this.location.getAuthority();
+			if (this.getBucketName() == null) {
+				throw new IllegalArgumentException("No bucket specified in the location: " + location);
+			}
+
+			// determine blob name
+			if (this.location.getPath() != null && this.location.getPath().length() > 1) {
+				this.blobName = this.location.getPath().substring(1);
+			}
+			else {
+				this.blobName = null;
+			}
+
+			// verify that if it's a bucket handle, location ends with a '/'
+			if (this.blobName == null && !location.endsWith("/")) {
+				throw new IllegalArgumentException("Invalid location for a bucket: " + location
```

Comment:
Could we handle these? I have a feeling it would be annoying as a user to have to keep this bit about the trailing slashes in mind.

Reply:
If the `/` is not there, relative URIs don't work right.

Also, I think it's better for the user to be explicit in whether they're declaring a bucket or a file path. It yields fewer surprises down the road.

Reply:
But couldn't you add it for them if it isn't there? 

Conceptually the two approaches are identical, but I think it's not too hard to have an if-it's-there-then-do-nothing/if-it's-not-there-then-add-it section.

Reply:
+1 for adding slashes for users.

Reply:
OK :)